features/netsocket/cellular/generic_modem_driver/PPPCellularInterface.cpp

```diff
@@ -565,7 +563,21 @@ nsapi_error_t PPPCellularInterface::connect()
     bool success;
     bool did_init = false;
     const char *apn_config = NULL;
-
+    bool user_specified_apn = false;
+
+    /* If the user has specified the APN then use that or,
+     * if we are not using the APN database, set _apn to
+     * "internet" as a best guess
+     */
+    if (_apn != NULL) {
+        user_specified_apn = true;
```

Comment:
This now has a different problem, in that this will have `user_specified_apn` true on the second connect with the database, because the _apn was set from the lookup. So the apn chosen for the first network seen becomes "sticky" and reused on the next network.

I think for clarity I would rename the members `user_apn`, etc. Then use local variables / parameters for the bringup, not those members directly. connect() should fill in local apn etc variables from lookup or copy them from the "user" members (or default to "internet").

setup_context_and_credentials should take "apn" as a parameter, not read the "user_apn" member.

Reply:
The APN you use is not based on the registered network, it is based on the IMSI of the SIM (i.e. the network providing your service, rather than the one you happen to be registered on), so unless the SIM is hot-swapped (which we have no concept of here) the choice will have already been made correctly (hence why it is inside the ` if (!initialized) {...` block just below).

I'll look into the other changes.

Reply:
Ah, okay, that changes things. Misunderstanding on my part. Given that, I guess this is reasonable, if we assume SIM swap means reboot.

Reply:
Yes, unless special arrangement have been made (very rare, and we don't have the code to deal with them here anyway), recognising a change of SIM requires a reboot.

Reply:
I tried the renaming/parameter passing as suggested but it ends up looking wrong to me since, in the case of database look-up, it is not really the user's APN, it is the default APN for the operator that provides their service.

There is a bigger issue, which I've been avoiding: `setup_context_and_credentials()` is only ever called once (inside the `if (initialized) {...` block), so I don't think it matters if a user passes a different APN into `connect(credentials)` on a subsequent call, that new APN would not be passed to the modem (with `AT+CGDCONT`) in any case.

Reply:
Right, okay, leave that change then. But if you could do the set_credentials inside connect at least.